[UNIT] llama-index framework - release 4/3/25 (#22)
huggingface/agents-course: units/en/unit2/llama-index/tools.mdx

1. `FunctionTool`: Convert any Python function into a tool that an agent can use. It automatically figures out how the function works.
2. `QueryEngineTool`: A tool that lets agents use query engines. Since agents are built on query engines, they can also use other agents as tools.
3. `Toolspecs`: Sets of tools created by the community, which often include tools for specific services like Gmail.
4. `Utility Tools`: Special tools that help handle large amounts of data from other tools.

We will go over each of them in more detail below.

## Creating a FunctionTool

A FunctionTool provides a simple way to wrap any Python function and make it available to an agent.
You can pass either a synchronous or asynchronous function to the tool, along with optional `name` and `description` parameters.
The name and description are particularly important as they help the agent understand when and how to use the tool effectively.
Let's look at how to create a FunctionTool below.

```python
from llama_index.core.tools import FunctionTool

def get_weather(location: str) -> str:
    """Usfeful for getting the weather for a given location."""
    ...

> tool = FunctionTool.from_defaults(
>     fn=get_weather,
>     # async_fn=aget_weather,
>     # name="...",
>     # description="...",
> )

Reviewer: The code will automatically detect async vs sync now

```suggestion
tool = FunctionTool.from_defaults(
    get_weather,
    # name="...",
    # description="...",
)
```

```

<Tip>When using an agent or LLM with function calling, the tool selected (and the arguments written for that tool) rely strongly on the tool name and description of the tools purpose and arguments. Learn more about function calling in the <a href="https://docs.llamaindex.ai/en/stable/module_guides/deploying/agents/modules/function_calling.html">Function Calling Guide</a> and <a href="https://docs.llamaindex.ai/en/stable/understanding/agent/function_calling.html">Function Calling Learning Guide</a>.</Tip>

## Creating a QueryEngineTool

The `QueryEngine` we defined in the previous unit can be easily transformed into a tool using the `QueryEngineTool` class.
Let's see how to create a `QueryEngineTool` from a `QueryEngine` in the example below.

```python

> from llama_index.core import StorageContext, load_index_from_storage
> from llama_index.core.tools import QueryEngineTool
> from llama_index.llms.huggingface_api import HuggingFaceInferenceAPILM
> from llama_index.embeddings.huggingface_api import HuggingFaceInferenceAPIEmbedding
> 
> embed_model = HuggingFaceInferenceAPIEmbedding("BAAI/bge-small-en-v1.5")
> storage_context = StorageContext.from_defaults(persist_dir="path/to/vector/store")
> index = load_index_from_storage(storage_context, embed_model=embed_model)

Reviewer: ```suggestion
from llama_index.core import VectorStoreIndex
from llama_index.core.tools import QueryEngineTool
from llama_index.llms.huggingface_api import HuggingFaceInferenceAPILM
from llama_index.embeddings.huggingface_api import HuggingFaceInferenceAPIEmbedding
from llama_index.vector_stores.chroma import ChromaVectorStore

embed_model = HuggingFaceInferenceAPIEmbedding("BAAI/bge-small-en-v1.5")

db = chromadb.PersistentClient(path="./alfed_chroma_db")
chroma_collection = db2.get_or_create_collection("alfred")
vector_store = ChromaVectorStore(chroma_collection=chroma_collection)

index = VectorStoreIndex.from_vector_store(vector_store, embed_model=embed_model)
```

llm = HuggingFaceInferenceAPILM(model_name="meta-llama/Meta-Llama-3-8B-Instruct")
query_engine = index.as_query_engine(llm=llm)

> tool = QueryEngineTool.from_defaults(query_engine)

Reviewer: ```suggestion
tool = QueryEngineTool.from_defaults(query_engine, name="some useful name", description="some useful description")
```

query_engine = index.as_query_engine(llm=llm)
tool = QueryEngineTool.from_defaults(query_engine, name="some useful name", description="some useful description")
```

## Creating Toolspecs

Think of `ToolSpecs` as collections of tools that work together harmoniously - like a well-organized professional toolkit.
Just as a mechanic's toolkit contains complementary tools that work together for vehicle repairs, a `ToolSpec` combines related tools for specific purposes.
For example, an accounting agent's `ToolSpec` might elegantly integrate spreadsheet capabilities, email functionality, and calculation tools to handle financial tasks with precision and efficiency.

<details>
<summary>Install the Google Toolspec</summary>
As introduced in the [section on components](what-are-components-in-llama-index.mdx), we can install the Google toolspec with the following command:

```python
pip install llama-index-tools-google
```
</details>

And now we can load the toolspec and convert it to a list of tools.

```python
from llama_index.tools.google import GmailToolSpec

tool_spec = GmailToolSpec()
tool_spec_list = tool_spec.to_tool_list()
```

## Utility Tools

> Oftentimes, directly querying an API can return an excessive amount of data, some of which may be irrelevant, overflow the context window of the LLM, or unnecessarily increase the number of tokens that you are using.

Reviewer 2: ```suggestion
Oftentimes, directly querying an API **can return an excessive amount of data**, some of which may be irrelevant, overflow the context window of the LLM, or unnecessarily increase the number of tokens that you are using.
```
